src/track/beats.h

```diff
@@ -51,121 +85,238 @@ class Beats : public QObject {
-    // capability BEATSCAP_ADDREMOVE.
-    virtual void addBeat(double dBeatSample) = 0;
+    /// Return the track signature at the given frame position
+    virtual TimeSignature getSignature(double sample = 0) const {
+        return getSignature(sample / 2.0);
+    }
+    virtual TimeSignature getSignatureNew(FrameNum frame = 0) const;
 
-    // Remove a beat at location dBeatSample. Beats instance must have the
+    /// Sets the nearest beat as a bar beat
+    virtual void setBar(double sample = 0) {
+        setBar(sample / 2.0);
+    }
+    virtual void setBarNew(FrameNum frame = 0);
+
+    /// Add a beat at location frame. Beats instance must have the
+    /// capability BEATSCAP_ADDREMOVE.
+    virtual void addBeat(double sample) {
+        addBeat(sample / 2.0);
+    }
+    virtual void addBeatNew(FrameNum frame);
+
+    // Remove a beat at location frame. Beats instance must have the
     // capability BEATSCAP_ADDREMOVE.
-    virtual void removeBeat(double dBeatSample) = 0;
+    virtual void removeBeat(double sample) {
+        removeBeatNew(sample / 2.0);
+    }
+    virtual void removeBeatNew(FrameNum frame);
 
+    // TODO(JVC) Do we want to move the beats a number of frames??
     // Translate all beats in the song by dNumSamples samples. Beats that lie
     // before the start of the track or after the end of the track are not
     // removed. Beats instance must have the capability BEATSCAP_TRANSLATE.
-    virtual void translate(double dNumSamples) = 0;
+    virtual void translate(double dNumSamples);
 
     // Scale the position of every beat in the song by dScalePercentage. Beats
     // class must have the capability BEATSCAP_SCALE.
-    virtual void scale(enum BPMScale scale) = 0;
+    virtual void scale(enum BPMScale scale);
 
     // Adjust the beats so the global average BPM matches dBpm. Beats class must
     // have the capability BEATSCAP_SET.
-    virtual void setBpm(double dBpm) = 0;
+    virtual void setBpm(double dBpm);
+
+    /// Returns the number of beats
+    inline int size() {
+        return m_beats.size();
+    }
+
+    /// Prints debuging information in stderr
+    void printDebugInfo() const;
+    /// Returen the frame number for the first beat, -1 is no beats
+    FrameNum getFirstBeatPosition() const;
+    /// Returns the frame number for the last beat, -1 if no beats
+    FrameNum getLastBeatPosition() const;
+
+  private:
+    void onBeatlistChanged();
+    bool readByteArray(const QByteArray& byteArray);
+    void scaleDouble();
+    void scaleTriple();
+    void scaleQuadruple();
+    void scaleHalve();
+    void scaleThird();
+    void scaleFourth();
+
+    mutable QMutex m_mutex;
+    const Track* m_track;
+    QString m_subVersion;
+    SINT m_iSampleRate;
+    double m_dCachedBpm;
+    FrameNum m_dLastFrame;
+    BeatList m_beats;
+    Logger m_logger;
 
   signals:
     void updated();
 };
 
-#endif /* BEATS_H */
+} // Namespace mixxx
```

Comment:
clang-format should produce this line:
} // namespace mixxx

Reply:
done